routers/api/v1/projects/project.go

```diff
@@ -0,0 +1,422 @@
+	// "404":
+	//    "$ref": "#/responses/notFound"
+	innerCreateProject(ctx, project_model.TypeOrganization)
+}
+
+func CreateRepoProject(ctx *context.APIContext) {
+	// swagger:operation POST /repos/{owner}/{repo}/projects project projectCreateRepositoryProject
+	// ---
+	// summary: Create a repository project
+	// produces:
+	// - application/json
+	// consumes:
+	// - application/json
+	// parameters:
+	//   - name: owner
+	//     in: path
+	//     description: owner of repo
+	//     type: string
+	//     required: true
+	//   - name: repo
+	//     in: path
+	//     description: repo
+	//     type: string
+	//     required: true
+	//   - name: project
+	//     in: body
+	//     required: true
+	//     schema: { "$ref": "#/definitions/NewProjectPayload" }
+	// responses:
+	//  "201":
+	//    "$ref": "#/responses/Project"
+	//  "403":
+	//    "$ref": "#/responses/forbidden"
+	//  "404":
+	//    "$ref": "#/responses/notFound"
+	innerCreateProject(ctx, project_model.TypeRepository)
+}
+
+func GetProject(ctx *context.APIContext) {
+	// swagger:operation GET /projects/{id} project projectGetProject
+	// ---
+	// summary: Get project
+	// produces:
+	// - application/json
+	// parameters:
+	//   - name: id
+	//     in: path
+	//     description: id of the project
+	//     type: string
+	//     required: true
+	// responses:
+	//  "200":
+	//    "$ref": "#/responses/Project"
+	//  "403":
+	//    "$ref": "#/responses/forbidden"
+	//  "404":
+	//    "$ref": "#/responses/notFound"
+	project, err := project_model.GetProjectByID(ctx, ctx.ParamsInt64(":id"))
+	if err != nil {
+		if project_model.IsErrProjectNotExist(err) {
+			ctx.NotFound()
+		} else {
+			ctx.Error(http.StatusInternalServerError, "GetProjectByID", err)
+		}
+		return
+	}
+
+	projectResponse, err := convert.ToAPIProject(ctx, project)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "GetProjectByID", err)
+		return
+	}
+	ctx.JSON(http.StatusOK, projectResponse)
+}
+
+func UpdateProject(ctx *context.APIContext) {
+	// swagger:operation PATCH /projects/{id} project projectUpdateProject
+	// ---
+	// summary: Update project
+	// produces:
+	// - application/json
+	// consumes:
+	// - application/json
+	// parameters:
+	//   - name: id
+	//     in: path
+	//     description: id of the project
+	//     type: string
+	//     required: true
+	//   - name: project
+	//     in: body
+	//     required: true
+	//     schema: { "$ref": "#/definitions/UpdateProjectPayload" }
+	// responses:
+	//  "200":
+	//    "$ref": "#/responses/Project"
+	//  "403":
+	//    "$ref": "#/responses/forbidden"
+	//  "404":
+	//    "$ref": "#/responses/notFound"
+	form := web.GetForm(ctx).(*api.UpdateProjectPayload)
+	project, err := project_model.GetProjectByID(ctx, ctx.ParamsInt64("id"))
+	if err != nil {
+		if project_model.IsErrProjectNotExist(err) {
+			ctx.NotFound()
+		} else {
+			ctx.Error(http.StatusInternalServerError, "UpdateProject", err)
+		}
+		return
+	}
+	if project.Title != form.Title {
```

Comment:
Shouldn't form Title and Description be *string and be updated only when provided? At least this is how normaly PATCH works. And I don't see any point in comparing with existing project values

Reply:
I agree with you on the first point to check if the title or description is present, 

As for the second point, the reason behind comparing with existing value is that if there is some logic somewhere down the code which reacts to the changes and if there is logic of diffing for changes then setting the value here blindly would cause unwanted side effects.

I might be thinking too much on this, but its practice i picked up from other places where this was an issue.